src/com/ms/silverking/cloud/dht/common/NamespaceOptionsClientBase.java

```diff
@@ -0,0 +1,547 @@
+
+    dhtName = dhtConfig.getName();
+    zkConfig = dhtConfig.getZKConfig();
+
+    if (debug) {
+      Log.warning("ensureNSCreationOptionsSet()");
+      Thread.dumpStack();
+    }
+    synchronized (this) {
+      if (nsCreationOptions == null) {
+        nsCreationOptions = nsCreationOptionsMap.get(dhtName);
+      }
+      if (nsCreationOptions == null) {
+        if (dhtMetaWatcher == null) {
+          assert nsCreationOptions == null;
+          try {
+            if (debug) {
+              Log.warning("ensureNSCreationOptionsSet() calling dhtMetaWatcher");
+            }
+            dhtMetaWatcher = new DHTMetaWatcher(zkConfig, dhtName, dhtMetaWatcherIntervalMillis, debug);
+            dhtMetaWatcher.waitForDHTConfiguration();
+            nsCreationOptions = dhtMetaWatcher.getDHTConfiguration().getNSCreationOptions();
+            nsCreationOptionsMap.put(dhtName, nsCreationOptions);
+            if (debug) {
+              Log.warning("nsCreationOptions: ", nsCreationOptions);
+            }
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        }
+      }
+    }
+  }
+
+  public NamespaceCreationOptions getNamespaceCreationOptions() {
+    ensureNSCreationOptionsSet();
+    return nsCreationOptions;
+  }
+
+  ///// ====== Internal client concurrency mechanism for nsOptions create/modify request ======
+  private final Lock activeOptionsRequestLock = new ReentrantLock();
+  private final HashedListMap<Long, NamespaceOptionsClientBase.ActiveOptionsRequest> activeOptionsRequests =
+      new HashedListMap<>(
+      true);
+  private static final int activeOptionsRequestTimeoutMillis = 2 * 60 * 1000;
+
+  private static class ActiveOptionsRequest {
+    private final long nsContext; // currently stored solely for debugging/logging
+    private final long absTimeoutMillis;
+    private NamespaceProperties storedData;
+
+    ActiveOptionsRequest(long nsContext, long relTimeoutMillis) {
+      this.nsContext = nsContext;
+      this.absTimeoutMillis = SystemTimeUtil.skSystemTimeSource.absTimeMillis() + relTimeoutMillis;
+    }
+
+    NamespaceProperties waitForCompletion() throws TimeoutException {
+      synchronized (this) {
+        while (storedData == null) {
+          long relDeadlineMillis;
+          long timeToWaitMillis;
+
+          relDeadlineMillis = Math.max(0, absTimeoutMillis - SystemTimeUtil.skSystemTimeSource.absTimeMillis());
+          try {
+            timeToWaitMillis = Math.min(activeOptionsRequestTimeoutMillis, relDeadlineMillis);
+            if (debug) {
+              System.out.printf("timeToWaitMillis %d\n", timeToWaitMillis);
+            }
+            if (timeToWaitMillis == 0) {
+              throw new TimeoutException(Long.toString(absTimeoutMillis));
+            }
+            this.wait(timeToWaitMillis);
+            if (debug) {
+              System.out.printf("out of wait\n");
+            }
+          } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)